### PR TITLE
3415 Update WSIReader

### LINKED ISSUE #3415: WSIReader with OpenSlide can load regions of incorrect size by default
**Describe the bug**
When `size` is unspecified and using the OpenSlide backend, `WSIReader.get_data()` assumes by default that the size of the slide at a given magnification `level` is `2**level` times smaller than the full-resolution size (see [here](https://github.com/Project-MONAI/MONAI/blob/0.8.0/monai/data/image_reader.py#L771-L773)). However, the magnification factors are not necessarily consecutive powers of 2, in which case the reader can return arrays of vastly incorrect size.

For example, I've encountered this issue with slides from the TCGA dataset. At the highest `level`, the image is downsampled by 32x instead of the assumed 8x, meaning the reader returns an image array that is 4x larger than necessary in each dimension (filled with 1/16th data, the rest padded with zeros).

**To Reproduce**
1. Download a pathology slide file with inhomogeneous downsampling levels (e.g. file "TCGA-2A-A8VL-01Z-00-DX1.2C2BD6EF-EC17-4117-AE89-A22B67AFB233.svs" from TCGA, available here on [the GDC portal](https://portal.gdc.cancer.gov/cases/49197847-cc83-4ce1-8397-d09cea4c4928?filters=%7B%22op%22%3A%22and%22%2C%22content%22%3A%5B%7B%22op%22%3A%22in%22%2C%22content%22%3A%7B%22field%22%3A%22cases.case_id%22%2C%22value%22%3A%5B%2249197847-cc83-4ce1-8397-d09cea4c4928%22%5D%7D%7D%5D%7D); has [1x, 4x, 16x, 32x] downsampling).
2. Install OpenSlide and its Python wrapper.
3. Instantiate `WSIReader(backend='openslide')`
```python
from monai.data.image_reader import WSIReader
reader = WSIReader(backend='openslide')
slide = reader.read("<your slide file>")
level = slide.level_count - 1  # pick highest level for demonstration
image = reader.get_data(slide, level=level)
# image is 4x4 times larger than expected and filled with 15/16=93.8% zeros
```

**Expected behavior**
The `WSIReader` should consider the actual magnification levels and corresponding dimensions encoded in the file headers. This information is readily available in the OpenSlide object's `level_dimensions` attribute ([docs](https://openslide.org/api/python/#openslide.OpenSlide.level_dimensions)).

**Environment**
- `monai` v0.6.0 (but same behaviour is still implemented in latest v0.8)
- `openslide` v3.4.1
- `openslide-python` v1.1.2

### PR DESCRIPTION
**Pull request by drbeh**
Fixes #3415 

### Description
This PR update WSIReader to evaluate level and size using metadata, improves patch extraction using `tifffile` backend, and adds new test case.

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests`.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @drbeh ,

Thanks for your quick update, does it fix the ticket: https://github.com/Project-MONAI/MONAI/issues/3415?

**Comment by drbeh** [comment]
@Nic-Ma do you know why the quick tests have failed?!

**Comment by Nic-Ma** [comment]
> @Nic-Ma do you know why the quick tests have failed?!

Let's try to run again, invite the user to verify first.

Thanks.

**Comment by drbeh** [comment]
Hi @dccastro, 
Could you please check if this PR solves the issue that you raised?
Thanks

CC @javier-alvarez

**Comment by dccastro** [comment]
Thanks for addressing this so quickly! I'll try this now and report back shortly.

**Comment by dccastro** [comment]
@drbeh @Nic-Ma I confirm this version now works on my test case. Thank you for looking into it 👍

**Comment by Nic-Ma** [comment]
I tried to run CI tests several times, seems something wrong with:
```
Ae_tiff (tests.test_wsireader.TestTiffFile) ... /home/runner/work/_temp/5daf4983-96e1-4862-927c-4b56b9f52ec5.sh: line 3:  1868 Killed
```
@drbeh @wyli do you have any idea?

Thanks in advance.

File: monai/data/image_reader.py [hunk 1]
```diff
@@ -788,30 +793,66 @@ def get_data(
 
         return patches, metadata
 
+    def _evaluate_level(self, img, level):
```
Comment (Nic-Ma) [inline monai/data/image_reader.py:796]: I feel the function name is a little bit confusing, maybe just call it `_check_level`?

Thanks.

File: monai/data/image_reader.py [hunk 2]
```diff
@@ -788,30 +793,66 @@ def get_data(
...
+            level_count = img.resolutions["level_count"]
+        elif self.backend == "tifffile":
+            level_count = len(img.pages)
+
+        if level > level_count - 1:
+            raise ValueError(f"The maximum level of this image is {level_count - 1} while level={level} is requested)!")
+
+        return level
+
+    def _evaluate_image_size(self, img, size, level, location):
```
Comment (Nic-Ma) [inline monai/data/image_reader.py:812]: Maybe call it `_get_image_size`?

Thanks.

**Comment by wyli** [comment]
> I tried to run CI tests several times, seems something wrong with:
> 
> ```
> Ae_tiff (tests.test_wsireader.TestTiffFile) ... /home/runner/work/_temp/5daf4983-96e1-4862-927c-4b56b9f52ec5.sh: line 3:  1868 Killed
> ```
> 
> @drbeh @wyli do you have any idea?
> 
> Thanks in advance.

maybe it's using too much CPU memory with the external IO package...

**Comment by drbeh** [comment]
> > I tried to run CI tests several times, seems something wrong with:
> > ```
> > Ae_tiff (tests.test_wsireader.TestTiffFile) ... /home/runner/work/_temp/5daf4983-96e1-4862-927c-4b56b9f52ec5.sh: line 3:  1868 Killed
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @drbeh @wyli do you have any idea?
> > Thanks in advance.
> 
> maybe it's using too much CPU memory with the external IO package...

That's right! Unlike cuCIM and OpenSlide, TiffFile load the entire image into memory to extract any region.
I skipped the test that loads image at level=0 (full resolution) for TiffFile backend.

File: tests/test_wsireader.py [hunk 3]
```diff
@@ -28,6 +28,8 @@
 has_cucim = has_cucim and hasattr(cucim, "CuImage")
 _, has_osl = optional_import("openslide")
 imsave, has_tiff = optional_import("tifffile", name="imsave")
+_, has_codec = optional_import("imagecodecs")
```
Comment (Nic-Ma) [inline tests/test_wsireader.py:31]: Should we add the installation of `imagecodecs` in the CI setup?

Thanks.

Comment (wyli) [reply tests/test_wsireader.py:31]: It's included https://github.com/Project-MONAI/MONAI/blob/f94e76824cf310d908e407394bda3b0507b3d3c0/requirements-dev.txt#L36

**Review by wyli — APPROVED** [review]

**Pull request merged**